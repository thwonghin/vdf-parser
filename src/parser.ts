import fs from 'node:fs';
import type stream from 'node:stream';

import {
    ControlType,
    TokenType,
    Tokenizer,
    type TokenResponse,
    type ControlResponse,
} from './tokenizer';
import { assertNever } from './utils';

export type VdfParsedKeyValue = {
    keys: string[];
    value: string;
};

export type VdfParserOptions = {
    disableEscape?: boolean;
    useLatestValue?: boolean;
    verbose?: boolean;
    debugBufferSize?: number;
};

export type VdfKeyValueMap = {
    [K: string]: string | VdfKeyValueMap;
};

export class VdfParserError extends Error {}

export class VdfParser {
    public readonly result: VdfKeyValueMap = {};

    private readonly keyStack: string[] = [];
    private readonly tokenizer: Tokenizer;

    constructor(private readonly options: VdfParserOptions = {}) {
        this.tokenizer = new Tokenizer(options);
    }

    /**
     * Parse a file and return an object
     * @example
     * ```
     * import { VdfParser } from '@hinw/vdf-parser';
     *
     * // file content: "key" { "nested_key" "value" }"
     * const filePath = 'input/sample.vdf';
     * const parser = new VdfParser();
     * const result = await parser.parseFile(filePath);
     *
     * // assert.assertEqual(result, { key: { nested_key: 'value' } });
     *
     * ```
     *
     * @param {string} filePath The path to the VDF file
     * @returns {VdfKeyValueMap} The parsed object
     */
    public async parseFile(filePath: string) {
        return this.parseStream(
            fs.createReadStream(filePath, { encoding: 'utf-8' }),
        );
    }

    /**
     * Parse a read stream and return an object
     * @example
     * ```
     * import { VdfParser } from '@hinw/vdf-parser';
     *
     * const readStream = stream.Readable.from(`"key" { "nested_key" "value" }"`);
     * const parser = new VdfParser();
     * const result = await parser.parseStream(readStream)
     *
     * // assert.assertEqual(result, { key: { nested_key: 'value' } });
     *
     * ```
     *
     * @param {stream.Readable} readStream The read stream contains the VDF content
     * @returns {VdfKeyValueMap} The parsed object
     */
    public async parseStream(
        readStream: stream.Readable,
    ): Promise<VdfKeyValueMap> {
        readStream.setEncoding('utf-8');
        for await (const chunk of readStream) {
            this.ingestText(chunk as string, true);
        }

        this.ingestChar('\n');
        this.flush();
        return this.result;
    }

    /**
     * Parse a text and return an object
     * @example
     * ```
     * import { VdfParser } from '@hinw/vdf-parser';
     *
     * const text = `"key" { "nested_key" "value" }"`;
     * const parser = new VdfParser();
     * const result = parser.parseText(text)
     *
     * // assert.assertEqual(result, { key: { nested_key: 'value' } });
     *
     * ```
     *
     * @param {string} text The VDF text content
     * @returns {VdfKeyValueMap} The parsed object
     */
    public parseText(
        text: string,
        skipAppendingNewline?: boolean,
    ): VdfKeyValueMap {
        this.ingestText(text, skipAppendingNewline);
        return this.result;
    }

    /**
     * Parse a file and return an async generator for the key value paris
     * @example
     * import { VdfParser } from '@hinw/vdf-parser';
     * // file content: "key" { "nested_key" "value" }"
     * const filePath = 'input/sample.vdf';
     *
     * const parser = new VdfParser();
     * const keyValuePairsIterator = parser.iterateKeyValuesFromFile(filePath);
     *
     * for await (const pair of keyValuePairsIterator) {
     *     // assert.assertEqual(pair, { keys: ['key', 'nested_key'], value: 'value' });
     * }
     *
     * // Or convert the generator as stream
     * import stream
     * const keyValuePairStream = stream.Readable.from(keyValuePairsIterator)
     *
     * @param {string} filePath The path to the VDF file
     * @returns {AsyncGenerator<VdfParsedKeyValue>} The async generator producing the key value pair
     */
    public async *iterateKeyValuesFromFile(
        filePath: string,
    ): AsyncGenerator<VdfParsedKeyValue> {
        yield* this.iterateKeyValuesFromReadStream(
            fs.createReadStream(filePath, { encoding: 'utf-8' }),
        );
    }

    /**
     * Parse a read stream and return an async generator for the key value paris
     * @example
     * import { VdfParser } from '@hinw/vdf-parser';
     *
     * const readStream = stream.Readable.from(`"key" { "nested_key" "value" }"`);
     * const parser = new VdfParser();
     * const keyValuePairsIterator = parser.iterateKeyValuesFromFile(filePath);
     *
     * for await (const pair of keyValuePairsIterator) {
     *     // assert.assertEqual(pair, { keys: ['key', 'nested_key'], value: 'value' });
     * }
     *
     * // Or convert the generator as stream
     * import stream
     * const keyValuePairStream = stream.Readable.from(keyValuePairsIterator)
     *
     * @param {stream.Readable} readStream The read stream contains the VDF content
     * @returns {AsyncGenerator<VdfParsedKeyValue>} The async generator producing the key value pair
     */
    public async *iterateKeyValuesFromReadStream(
        readStream: stream.Readable,
    ): AsyncGenerator<VdfParsedKeyValue> {
        readStream.setEncoding('utf-8');
        for await (const chunk of readStream) {
            for (const c of chunk) {
                yield* this.iterateKeyValues(c as string);
            }
        }

        yield* this.iterateKeyValues('\n');
        yield* this.flushTokenizer();
    }

    private ingestChar(char: string) {
        for (const pair of this.iterateKeyValues(char)) {
            this.buildKvMap(pair);
        }
    }

    private ingestText(text: string, skipAppendingNewline?: boolean) {
        for (const char of text) {
            this.ingestChar(char);
        }

        if (!skipAppendingNewline) {
            this.ingestChar('\n');
        }
    }

    private flush() {
        for (const pair of this.flushTokenizer()) {
            this.buildKvMap(pair);
        }
    }

    private *iterateKeyValues(char: string) {
        if ([...char].length !== 1) {
            throw new VdfParserError(
                'Should ingest 1 character each time. Use `ingestText` for multiple characters',
            );
        }

        for (const response of this.tokenizer.ingestChar(char)) {
            yield* this.parseTokenResponse(response);
        }
    }

    private *flushTokenizer() {
        for (const response of this.tokenizer.flush()) {
            yield* this.parseTokenResponse(response);
        }
    }

    private *parseTokenResponse(response: TokenResponse | ControlResponse) {
        if ('tokenType' in response) {
            switch (response.tokenType) {
                case TokenType.KEY:
                    this.keyStack.push(response.token);
                    break;
                case TokenType.VALUE:
                    yield {
                        keys: [...this.keyStack],
                        value: response.token,
                    };
                    this.keyStack.pop();
                    break;
                default:
                    assertNever(response.tokenType);
            }
        }

        if ('controlType' in response) {
            switch (response.controlType) {
                case ControlType.START_NESTED:
                    // Nothing need to be done
                    break;
                case ControlType.END_NESTED:
                    this.keyStack.pop();
                    break;
                default:
                    assertNever(response.controlType);
            }
        }
    }

    private buildKvMap(pair: VdfParsedKeyValue) {
        const lastKey = pair.keys.pop();
        if (!lastKey) {
            throw new VdfParserError('Empty key encountered');
        }

        let traversed: VdfKeyValueMap = this.result;
        let key;

        while ((key = pair.keys.shift())) {
            const tempTraversed = traversed[key];

            if (tempTraversed === undefined) {
                const newKvMap = {};
                traversed[key] = newKvMap;
                traversed = newKvMap;
            } else if (typeof tempTraversed === 'string') {
                if (this.options.useLatestValue) {
                    const newKvMap = {};
                    traversed[key] = newKvMap;
                    traversed = newKvMap;
                } else {
                    return;
                }
            } else {
                traversed = tempTraversed;
            }
        }

        if (traversed[lastKey] === undefined || this.options.useLatestValue) {
            traversed[lastKey] = pair.value;
        }
    }
}
