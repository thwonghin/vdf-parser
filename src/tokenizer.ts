import { assertNever } from './utils';

enum TokenizerState {
    IN_KEY_WITH_QUOTE,
    IN_VALUE_WITH_QUOTE,
    IN_KEY_WITHOUT_QUOTE,
    IN_VALUE_WITHOUT_QUOTE,
    AFTER_KEY,
    AFTER_VALUE,
    COMMENT_AFTER_KEY,
    COMMENT_AFTER_VALUE,
}

const ESCAPE = '\\';
const ESCAPE_SEQUENCES__KEEP_SLASH = ['\\n', '\\t'];
const ESCAPE_SEQUENCES__NO_SLASH = ['\\\\', '\\"'];
const QUOTE = '"';
const WHITE_SPACE = ' ';
const TAB_SPACE = '\t';
const BOM = String.fromCharCode(65279);
const NEW_LINE_CHAR = '\n';
const CARRIAGE_RETURN_CHAR = '\r';
const BRACKET_OPEN = '{';
const BRACKET_CLOSE = '}';
const COMMENT_START = '//';

const DEFAULT_DEBUG_BUFFER_SIZE = 20;

export enum TokenType {
    KEY,
    VALUE,
    NEST,
}

export enum NestDirection {
    START = '{',
    END = '}',
}

export type TokenResponse = { tokenType: TokenType; value: string };

export type TokenizerOption = {
    disableEscape?: boolean;
    verbose?: boolean;
    debugBufferSize?: number;
};

export class TokenizerError extends Error {
    constructor(message: string, line: number, column: number, near: string) {
        super(
            `At line ${line}:${column}${near ? ` near ${near}` : ''}: ${message}`,
        );
    }
}
export class TokenizerTooManyBracketsError extends TokenizerError {
    constructor(line: number, column: number, near: string) {
        super('Too many closing brackets.', line, column, near);
    }
}
export class TokenizerOpenBracketAfterValueError extends TokenizerError {
    constructor(line: number, column: number, near: string) {
        super('Unexpected open bracket after value.', line, column, near);
    }
}
export class TokenizerCloseBracketAfterKeyError extends TokenizerError {
    constructor(line: number, column: number, near: string) {
        super('Unexpected close bracket after key.', line, column, near);
    }
}
export class TokenizerNoCharacterToEscapeError extends TokenizerError {
    constructor(line: number, column: number, near: string) {
        super('No character to escape.', line, column, near);
    }
}
export class TokenizerUnsupportedEscapeSequenceError extends TokenizerError {
    constructor(sequence: string, line: number, column: number, near: string) {
        super(`Unsupported escape sequence: ${sequence}.`, line, column, near);
    }
}
export class TokenizerEscapeOutsideQuote extends TokenizerError {
    constructor(line: number, column: number, near: string) {
        super('Cannot escape outside of quoted key/value.', line, column, near);
    }
}

// https://developer.valvesoftware.com/wiki/KeyValues#About_KeyValues_Text_File_Format
export class Tokenizer {
    private state: TokenizerState = TokenizerState.AFTER_VALUE;
    private readonly tokenParts: string[] = [];
    private nestedLevel = 0;
    private buffer: string | null = null;
    private currentLineNumber = 0;
    private currentPosition = 0;
    private readonly debugBuffer: string[] = [];

    constructor(private readonly options: TokenizerOption = {}) {}

    public *ingestChar(char: string): Generator<TokenResponse> {
        if (this.buffer === null) {
            this.buffer = char;
            return;
        }

        yield* this.parseCharacter(this.buffer, char);
        if (this.buffer !== null) {
            this.buffer = char;
        }

        this.storeDebugBuffer(char);
    }

    public *flush(): Generator<TokenResponse> {
        yield* this.ingestChar(NEW_LINE_CHAR);
        if (this.buffer !== null) {
            yield* this.parseCharacter(this.buffer, undefined);
            this.buffer = null;
        }
    }

    private storeDebugBuffer(char: string) {
        if (!this.options.verbose) {
            return;
        }

        this.debugBuffer.push(char);
        this.debugBuffer.splice(
            0,
            this.debugBuffer.length -
                (this.options.debugBufferSize ?? DEFAULT_DEBUG_BUFFER_SIZE),
        );
    }

    private *parseCharacter(
        char: string,
        lookahead: string | undefined,
    ): Generator<TokenResponse> {
        switch (char) {
            case CARRIAGE_RETURN_CHAR:
            case BOM:
                // Ignore. Do nothing
                break;
            case NEW_LINE_CHAR:
                yield* this.handleNewLine(char);
                break;
            case WHITE_SPACE:
            case TAB_SPACE:
                yield* this.handleWhitespace(char);
                break;
            case QUOTE:
                yield* this.handleQuote();
                break;
            case BRACKET_OPEN:
                yield* this.handleBracketOpen(char);
                break;
            case BRACKET_CLOSE:
                yield* this.handleBracketClose(char);
                break;
            case ESCAPE:
                if (this.options.disableEscape) {
                    this.handleNormalCharacter(char);
                } else {
                    this.handleEscape(char, lookahead);
                    this.buffer = null;
                }

                break;
            case COMMENT_START[0]:
                if ([char, lookahead].join('') === COMMENT_START) {
                    yield* this.handleComment();
                    this.buffer = null;
                } else {
                    this.handleNormalCharacter(char);
                }

                break;
            default:
                this.handleNormalCharacter(char);
        }

        if (this.options.verbose) {
            console.log(
                `Parsed ${this.currentLineNumber}:${this.currentPosition} ${char === NEW_LINE_CHAR ? 'new_line' : char} ${char.charCodeAt(0)}, State: ${this.state}`,
            );
        }
    }

    private *handleNewLine(char: string) {
        switch (this.state) {
            case TokenizerState.IN_KEY_WITH_QUOTE:
                this.tokenParts.push(char);
                break;
            case TokenizerState.IN_VALUE_WITH_QUOTE:
                this.tokenParts.push(char);
                break;
            case TokenizerState.IN_KEY_WITHOUT_QUOTE:
                yield this.emitToken(TokenType.KEY);
                this.state = TokenizerState.AFTER_KEY;
                break;
            case TokenizerState.IN_VALUE_WITHOUT_QUOTE:
                yield this.emitToken(TokenType.VALUE);
                this.state = TokenizerState.AFTER_VALUE;
                break;
            case TokenizerState.AFTER_KEY:
                // Do nothing
                break;
            case TokenizerState.AFTER_VALUE:
                // Do nothing
                break;
            case TokenizerState.COMMENT_AFTER_KEY:
                this.state = TokenizerState.AFTER_KEY;
                break;
            case TokenizerState.COMMENT_AFTER_VALUE:
                this.state = TokenizerState.AFTER_VALUE;
                break;
            default:
                assertNever(this.state);
        }

        this.currentLineNumber++;
        this.currentPosition = 0;
    }

    private *handleWhitespace(char: string) {
        switch (this.state) {
            case TokenizerState.IN_KEY_WITH_QUOTE:
                this.tokenParts.push(char);
                break;
            case TokenizerState.IN_VALUE_WITH_QUOTE:
                this.tokenParts.push(char);
                break;
            case TokenizerState.IN_KEY_WITHOUT_QUOTE:
                yield this.emitToken(TokenType.KEY);
                this.state = TokenizerState.AFTER_KEY;
                break;
            case TokenizerState.IN_VALUE_WITHOUT_QUOTE:
                yield this.emitToken(TokenType.VALUE);
                this.state = TokenizerState.AFTER_VALUE;
                break;
            case TokenizerState.AFTER_KEY:
                // Do nothing
                break;
            case TokenizerState.AFTER_VALUE:
                // Do nothing
                break;
            case TokenizerState.COMMENT_AFTER_KEY:
                // Do nothing
                break;
            case TokenizerState.COMMENT_AFTER_VALUE:
                // Do nothing
                break;
            default:
                assertNever(this.state);
        }

        this.currentPosition++;
    }

    private *handleQuote() {
        switch (this.state) {
            case TokenizerState.IN_KEY_WITH_QUOTE:
                yield this.emitToken(TokenType.KEY);
                this.state = TokenizerState.AFTER_KEY;
                break;
            case TokenizerState.IN_VALUE_WITH_QUOTE:
                yield this.emitToken(TokenType.VALUE);
                this.state = TokenizerState.AFTER_VALUE;
                break;
            case TokenizerState.IN_KEY_WITHOUT_QUOTE:
                yield this.emitToken(TokenType.KEY);
                this.state = TokenizerState.AFTER_KEY;
                break;
            case TokenizerState.IN_VALUE_WITHOUT_QUOTE:
                yield this.emitToken(TokenType.VALUE);
                this.state = TokenizerState.AFTER_VALUE;
                break;
            case TokenizerState.AFTER_KEY:
                this.state = TokenizerState.IN_VALUE_WITH_QUOTE;
                break;
            case TokenizerState.AFTER_VALUE:
                this.state = TokenizerState.IN_KEY_WITH_QUOTE;
                break;
            case TokenizerState.COMMENT_AFTER_KEY:
                // Do nothing
                break;
            case TokenizerState.COMMENT_AFTER_VALUE:
                // Do nothing
                break;
            default:
                assertNever(this.state);
        }

        this.currentPosition++;
    }

    private *handleBracketOpen(char: string) {
        switch (this.state) {
            case TokenizerState.IN_KEY_WITH_QUOTE:
                this.tokenParts.push(char);
                break;
            case TokenizerState.IN_VALUE_WITH_QUOTE:
                this.tokenParts.push(char);
                break;
            case TokenizerState.IN_KEY_WITHOUT_QUOTE:
                yield this.emitToken(TokenType.KEY);
                yield this.emitNest(NestDirection.START);
                this.state = TokenizerState.AFTER_VALUE;
                break;
            case TokenizerState.IN_VALUE_WITHOUT_QUOTE:
                throw new TokenizerOpenBracketAfterValueError(
                    this.currentLineNumber,
                    this.currentPosition,
                    this.debugBuffer.join(''),
                );
            case TokenizerState.AFTER_KEY:
                yield this.emitNest(NestDirection.START);
                this.state = TokenizerState.AFTER_VALUE;
                break;
            case TokenizerState.AFTER_VALUE:
                throw new TokenizerOpenBracketAfterValueError(
                    this.currentLineNumber,
                    this.currentPosition,
                    this.debugBuffer.join(''),
                );
            case TokenizerState.COMMENT_AFTER_KEY:
                // Do nothing
                break;
            case TokenizerState.COMMENT_AFTER_VALUE:
                // Do nothing
                break;
            default:
                assertNever(this.state);
        }

        this.currentPosition++;
    }

    private *handleBracketClose(char: string) {
        switch (this.state) {
            case TokenizerState.IN_KEY_WITH_QUOTE:
                this.tokenParts.push(char);
                break;
            case TokenizerState.IN_VALUE_WITH_QUOTE:
                this.tokenParts.push(char);
                break;
            case TokenizerState.IN_KEY_WITHOUT_QUOTE:
                throw new TokenizerCloseBracketAfterKeyError(
                    this.currentLineNumber,
                    this.currentPosition,
                    this.debugBuffer.join(''),
                );
            case TokenizerState.IN_VALUE_WITHOUT_QUOTE:
                yield this.emitToken(TokenType.VALUE);
                yield this.emitNest(NestDirection.END);
                this.state = TokenizerState.AFTER_VALUE;
                break;
            case TokenizerState.AFTER_KEY:
                throw new TokenizerCloseBracketAfterKeyError(
                    this.currentLineNumber,
                    this.currentPosition,
                    this.debugBuffer.join(''),
                );
            case TokenizerState.AFTER_VALUE:
                yield this.emitNest(NestDirection.END);
                break;
            case TokenizerState.COMMENT_AFTER_KEY:
                // Do nothing
                break;
            case TokenizerState.COMMENT_AFTER_VALUE:
                // Do nothing
                break;
            default:
                assertNever(this.state);
        }

        this.currentPosition++;
    }

    private handleEscape(char: string, lookAhead: string | undefined) {
        if (
            this.state === TokenizerState.COMMENT_AFTER_KEY ||
            this.state === TokenizerState.COMMENT_AFTER_VALUE
        ) {
            this.currentPosition++;
            // Do nothing
            return;
        }

        if (lookAhead === '' || lookAhead === undefined) {
            throw new TokenizerNoCharacterToEscapeError(
                this.currentLineNumber,
                this.currentPosition,
                this.debugBuffer.join(''),
            );
        }

        const charWithLookAhead = [char, lookAhead].join('');

        const keepSlash =
            ESCAPE_SEQUENCES__KEEP_SLASH.includes(charWithLookAhead);
        const removeSlash =
            ESCAPE_SEQUENCES__NO_SLASH.includes(charWithLookAhead);

        if (keepSlash === removeSlash) {
            // Unsupported slash sequence - handle like normal
            this.handleNormalCharacter(char);
            this.handleNormalCharacter(lookAhead);
            this.currentPosition += 2;
            return;
        }

        const charToPush = keepSlash ? charWithLookAhead : lookAhead;

        switch (this.state) {
            case TokenizerState.IN_KEY_WITH_QUOTE:
                this.tokenParts.push(charToPush);
                break;
            case TokenizerState.IN_VALUE_WITH_QUOTE:
                this.tokenParts.push(charToPush);
                break;
            case TokenizerState.IN_KEY_WITHOUT_QUOTE:
                throw new TokenizerEscapeOutsideQuote(
                    this.currentLineNumber,
                    this.currentPosition,
                    this.debugBuffer.join(''),
                );
            case TokenizerState.IN_VALUE_WITHOUT_QUOTE:
                throw new TokenizerEscapeOutsideQuote(
                    this.currentLineNumber,
                    this.currentPosition,
                    this.debugBuffer.join(''),
                );
            case TokenizerState.AFTER_KEY:
                throw new TokenizerEscapeOutsideQuote(
                    this.currentLineNumber,
                    this.currentPosition,
                    this.debugBuffer.join(''),
                );
            case TokenizerState.AFTER_VALUE:
                throw new TokenizerEscapeOutsideQuote(
                    this.currentLineNumber,
                    this.currentPosition,
                    this.debugBuffer.join(''),
                );
            default:
                assertNever(this.state);
        }

        this.currentPosition += 2;
    }

    private handleNormalCharacter(char: string) {
        switch (this.state) {
            case TokenizerState.IN_KEY_WITH_QUOTE:
                this.tokenParts.push(char);
                break;
            case TokenizerState.IN_VALUE_WITH_QUOTE:
                this.tokenParts.push(char);
                break;
            case TokenizerState.IN_KEY_WITHOUT_QUOTE:
                this.tokenParts.push(char);
                break;
            case TokenizerState.IN_VALUE_WITHOUT_QUOTE:
                this.tokenParts.push(char);
                break;
            case TokenizerState.AFTER_KEY:
                this.tokenParts.push(char);
                this.state = TokenizerState.IN_VALUE_WITHOUT_QUOTE;
                break;
            case TokenizerState.AFTER_VALUE:
                this.tokenParts.push(char);
                this.state = TokenizerState.IN_KEY_WITHOUT_QUOTE;
                break;
            case TokenizerState.COMMENT_AFTER_KEY:
                // Do nothing
                break;
            case TokenizerState.COMMENT_AFTER_VALUE:
                // Do nothing
                break;
            default:
                assertNever(this.state);
        }

        this.currentPosition++;
    }

    private *handleComment() {
        switch (this.state) {
            case TokenizerState.IN_KEY_WITH_QUOTE:
                yield this.emitToken(TokenType.KEY);
                this.state = TokenizerState.COMMENT_AFTER_KEY;
                break;
            case TokenizerState.IN_VALUE_WITH_QUOTE:
                yield this.emitToken(TokenType.VALUE);
                this.state = TokenizerState.COMMENT_AFTER_VALUE;
                break;
            case TokenizerState.IN_KEY_WITHOUT_QUOTE:
                yield this.emitToken(TokenType.KEY);
                this.state = TokenizerState.COMMENT_AFTER_KEY;
                break;
            case TokenizerState.IN_VALUE_WITHOUT_QUOTE:
                yield this.emitToken(TokenType.VALUE);
                this.state = TokenizerState.COMMENT_AFTER_VALUE;
                break;
            case TokenizerState.AFTER_KEY:
                this.state = TokenizerState.COMMENT_AFTER_KEY;
                break;
            case TokenizerState.AFTER_VALUE:
                this.state = TokenizerState.COMMENT_AFTER_VALUE;
                break;
            case TokenizerState.COMMENT_AFTER_KEY:
                // Do nothing
                break;
            case TokenizerState.COMMENT_AFTER_VALUE:
                // Do nothing
                break;
            default:
                assertNever(this.state);
        }

        this.currentPosition++;
    }

    private emitToken(tokenType: TokenType): TokenResponse {
        const result = { tokenType, value: this.tokenParts.join('') };
        this.tokenParts.splice(0);
        return result;
    }

    private emitNest(nestDirection: NestDirection) {
        switch (nestDirection) {
            case NestDirection.START:
                this.nestedLevel++;
                break;
            case NestDirection.END:
                this.nestedLevel--;
                if (this.nestedLevel < 0) {
                    throw new TokenizerTooManyBracketsError(
                        this.currentLineNumber,
                        this.currentPosition,
                        this.debugBuffer.join(''),
                    );
                }

                break;
            default:
                assertNever(nestDirection);
        }

        return { tokenType: TokenType.NEST, value: nestDirection };
    }
}
