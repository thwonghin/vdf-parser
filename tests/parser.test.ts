import fs from 'node:fs';
import path from 'node:path';
import stream from 'node:stream';

import { test, expect, describe } from 'bun:test';

import { VdfParser } from '../src/parser';
import { TokenizerError } from '../src/tokenizer';

const FIXTURE_OBJ_RESULT = {
    key: {
        'ke"y2': {
            key3: 'val\\ue3',
        },
    },
    key4: {
        none: 'none',
    },
    key5: {
        'key\\n{6}': 'val\\tue{6}',
        key7: {
            key8: 'value8',
        },
        key9: 'Ð¤ÑƒÐ»Ð° (ðž¤Šðž¤µðž¤¤ðž¤¬ðž¤µðž¤¤ðž¤£ðž¤«)',
        ä½ å¥½: 'ä¸–ç•Œ',
    },
};

const FIXTURE_PAIRS_RESULT = [
    {
        keyParts: ['key', 'ke"y2', 'key3'],
        value: 'val\\ue3',
    },
    {
        keyParts: ['key4', 'none'],
        value: 'none',
    },
    {
        keyParts: ['key4'],
        value: 'value4',
    },
    {
        keyParts: ['key5', 'key\\n{6}'],
        value: 'val\\tue{6}',
    },
    {
        keyParts: ['key5', 'key7', 'key8'],
        value: 'value8',
    },
    {
        keyParts: ['key5', 'key7', 'key8', 'key9'],
        value: 'value9',
    },
    {
        keyParts: ['key5', 'key9'],
        value: 'Ð¤ÑƒÐ»Ð° (ðž¤Šðž¤µðž¤¤ðž¤¬ðž¤µðž¤¤ðž¤£ðž¤«)',
    },
    {
        keyParts: ['key5', 'ä½ å¥½'],
        value: 'ä¸–ç•Œ',
    },
];

describe('parseText', () => {
    test('should parse correctly with escape', async () => {
        const input = `key{"ke\\"y2"{key3 "val\\\\ue3"}
        }
        key4 value4 // comment {} "" \\\\
        key5 {
            "key\\n{6}" "val\\tue{6}"
            "key7" {
                key8 "value8"
            }
            key9 "value9"
        }
        `;

        const parser = new VdfParser({ useLatestValue: true });
        const result = await parser.parseText(input);

        expect(result).toEqual({
            key: {
                'ke"y2': {
                    key3: 'val\\ue3',
                },
            },
            key4: 'value4',
            key5: {
                'key\\n{6}': 'val\\tue{6}',
                key7: {
                    key8: 'value8',
                },
                key9: 'value9',
            },
        });
    });

    test('should parse correctly without escape', async () => {
        const input = `key{"ke\\y2"{key3 "val\\\\ue3"}
        }
        key4 value4 // comment {} "" \\\\
        key5 {
            "key\\n{6}" "val\\tue{6}"
            "key7" {
                key8 "value8"
            }
            key9 "value9"
        }
        `;

        const parser = new VdfParser({
            disableEscape: true,
            useLatestValue: true,
        });
        const result = await parser.parseText(input);

        expect(result).toEqual({
            key: {
                'ke\\y2': {
                    key3: 'val\\\\ue3',
                },
            },
            key4: 'value4',
            key5: {
                'key\\n{6}': 'val\\tue{6}',
                key7: {
                    key8: 'value8',
                },
                key9: 'value9',
            },
        });
    });

    test('should parse correctly with latest value', async () => {
        const input = `key{"ke\\"y2"{key3 "val\\\\ue3"}
        }
        key4 {
            none none
        }
        key4 value4 // comment {} "" \\\\
        key5 {
            "key\\n{6}" "val\\tue{6}"
            "key7" {
                key8 "value8"
                key8 {
                    key9 "value9"
                }
            }
            key9 "value9"
        }
        `;

        const parser = new VdfParser({ useLatestValue: true });
        const result = await parser.parseText(input);

        expect(result).toEqual({
            key: {
                'ke"y2': {
                    key3: 'val\\ue3',
                },
            },
            key4: 'value4',
            key5: {
                'key\\n{6}': 'val\\tue{6}',
                key7: {
                    key8: {
                        key9: 'value9',
                    },
                },
                key9: 'value9',
            },
        });
    });

    test('should parse correctly without using latest value', async () => {
        const input = `key{"ke\\"y2"{key3 "val\\\\ue3"}
        }
        key4 {
            none none
        }
        key4 value4 // comment {} "" \\\\
        key5 {
            "key\\n{6}" "val\\tue{6}"
            "key7" {
                key8 "value8"
                key8 {
                    key9 "value9"
                }
            }
            key9 "value9"
        }
        `;

        const parser = new VdfParser();
        const result = await parser.parseText(input);

        expect(result).toEqual({
            key: {
                'ke"y2': {
                    key3: 'val\\ue3',
                },
            },
            key4: {
                none: 'none',
            },
            key5: {
                'key\\n{6}': 'val\\tue{6}',
                key7: {
                    key8: 'value8',
                },
                key9: 'value9',
            },
        });
    });

    test('should throw tokenizer error if malformed', async () => {
        const input = `key {}}`;

        const parser = new VdfParser();
        let error;
        try {
            await parser.parseText(input);
        } catch (err) {
            error = err;
        }

        expect(error instanceof TokenizerError).toBeTrue();
    });
});

describe('parseStream', () => {
    test('should parse correctly', async () => {
        const input = `key{"ke\\"y2"{key3 "val\\\\ue3"}
        }
        key4 {
            none none
        }
        key4 value4 // comment {} "" \\\\
        key5 {
            "key\\n{6}" "val\\tue{6}"
            "key7" {
                key8 "value8"
                key8 {
                    key9 "value9"
                }
            }
            key9 "value9"
        }
        `;

        const parser = new VdfParser();
        const result = await parser.parseStream(stream.Readable.from(input));

        expect(result).toEqual({
            key: {
                'ke"y2': {
                    key3: 'val\\ue3',
                },
            },
            key4: {
                none: 'none',
            },
            key5: {
                'key\\n{6}': 'val\\tue{6}',
                key7: {
                    key8: 'value8',
                },
                key9: 'value9',
            },
        });
    });
});

describe('parseFile', () => {
    test('should parse correctly', async () => {
        const parser = new VdfParser();
        const result = await parser.parseFile(
            path.join(__dirname, 'fixtures', 'sample.vdf'),
        );

        expect(result).toEqual(FIXTURE_OBJ_RESULT);
    });
});

describe('condensePairsAsync', () => {
    test('should return result correctly', async () => {
        const parser = new VdfParser();

        const readStream = fs.createReadStream(
            path.join(__dirname, 'fixtures', 'sample.vdf'),
            { encoding: 'utf-8' },
        );

        const result = await parser.condensePairsAsync(readStream.pipe(parser));

        expect(result).toEqual(FIXTURE_OBJ_RESULT);
    });
});

describe('condensePairs', () => {
    test('should return result correctly', async () => {
        const parser = new VdfParser();

        const readStream = fs.createReadStream(
            path.join(__dirname, 'fixtures', 'sample.vdf'),
            { encoding: 'utf-8' },
        );

        const paris = await Array.fromAsync(readStream.pipe(parser));

        const result = parser.condensePairs(paris);

        expect(result).toEqual(FIXTURE_OBJ_RESULT);
    });
});

describe('stream pipe', () => {
    test('should return key value pairs correctly', async () => {
        const parser = new VdfParser();

        const readStream = fs.createReadStream(
            path.join(__dirname, 'fixtures', 'sample.vdf'),
            { encoding: 'utf-8' },
        );

        const result = await Array.fromAsync(readStream.pipe(parser));

        expect(result).toEqual(FIXTURE_PAIRS_RESULT);
    });
});
