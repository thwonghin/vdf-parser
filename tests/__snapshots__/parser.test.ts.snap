// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`iterateKeyValuesFromReadStream should iterate correctly 1`] = `
[
  {
    "keys": [
      "key",
      "ke"y2",
      "key3",
    ],
    "value": "val\ue3",
  },
  {
    "keys": [
      "key4",
      "none",
    ],
    "value": "none",
  },
  {
    "keys": [
      "key4",
    ],
    "value": "value4",
  },
  {
    "keys": [
      "key5",
      "key\n{6}",
    ],
    "value": "val\tue{6}",
  },
  {
    "keys": [
      "key5",
      "key7",
      "key8",
    ],
    "value": "value8",
  },
  {
    "keys": [
      "key5",
      "key7",
      "key8",
      "key9",
    ],
    "value": "value9",
  },
  {
    "keys": [
      "key5",
      "key9",
    ],
    "value": "value9",
  },
]
`;

exports[`iterateKeyValuesFromFile should iterate correctly 1`] = `
[
  {
    "keys": [
      "key",
      "ke"y2",
      "key3",
    ],
    "value": "val\ue3",
  },
  {
    "keys": [
      "key4",
      "none",
    ],
    "value": "none",
  },
  {
    "keys": [
      "key4",
    ],
    "value": "value4",
  },
  {
    "keys": [
      "key5",
      "key\n{6}",
    ],
    "value": "val\tue{6}",
  },
  {
    "keys": [
      "key5",
      "key7",
      "key8",
    ],
    "value": "value8",
  },
  {
    "keys": [
      "key5",
      "key7",
      "key8",
      "key9",
    ],
    "value": "value9",
  },
  {
    "keys": [
      "key5",
      "key9",
    ],
    "value": "value9",
  },
  {
    "keys": [
      "key5",
      "你好",
    ],
    "value": "世界",
  },
]
`;
